rasa/nlu/data_augmentation/augmenter.py

```diff
@@ -0,0 +1,623 @@
+    augmented_training_data = _build_augmented_training_data(
+        nlu_training_data=nlu_training_data,
+        training_data_pool=training_data_pool,
+        data_augmentation_pool=random_expansion,
+    )
+
+    # Store augmented training data to file
+    rasa.shared.utils.io.create_directory(output_directory)
+    nlu_training_file = os.path.join(output_directory, "train_augmented_random.yml")
+    augmented_training_data.persist_nlu(filename=nlu_training_file)
+
+    # Run NLU train/test loop
+    intent_report = _train_test_nlu_model(
+        output_directory=output_directory,
+        nlu_training_file=nlu_training_file,
+        config=config,
+        nlu_evaluation_data=nlu_evaluation_data,
+    )
+
+    # Create data augmentation summary
+    (intent_summary, _) = _create_summary_report(
+        intent_report=intent_report,
+        classification_report=classification_report,
+        training_intents=nlu_training_data.intents,
+        pooled_intents=pooled_intents,
+        output_directory=output_directory,
+    )
+
+    # Plot data augmentation summary
+    _plot_summary_report(
+        intent_summary=intent_summary, output_directory=output_directory,
+    )
+
+
+def augment_nlu_data(
+    nlu_training_data: TrainingData,
+    nlu_evaluation_data: TrainingData,
+    paraphrases: TrainingData,
+    classification_report: Dict[Text, Dict[Text, float]],
+    config: Text,
+    num_intents_to_augment: int,
+    random_seed: int,
+    paraphrase_sim_score_threshold: float,
+    output_directory: Text,
+) -> None:
+    """Performs data augmentation for NLU by evaluating two augmentation strategies.
+
+    Args:
+        nlu_training_data: NLU training data (without data augmentation).
+        nlu_evaluation_data: NLU evaluation data.
+        paraphrases: The generated paraphrases with similarity scores obtained from https://github.com/RasaHQ/paraphraser.
+        classification_report: Classification report of the model run *without* data augmentation.
+        config: NLU model config.
+        num_intents_to_augment: Number of intents to choose for augmentation (per criterion).
+        random_seed: Random seed for sampling the paraphrases.
+        paraphrase_sim_score_threshold: Minimum required similarity for a generated paraphrase to be considered for data augmentation.
+        output_directory: Directory to store the output files in.
+    """
+    # Determine intents for which to perform data augmentation
+    pooled_intents = collect_intents_for_data_augmentation(
+        nlu_training_data=nlu_training_data,
+        num_intents_to_augment=num_intents_to_augment,
+        classification_report=classification_report,
+    )
+
+    # Retrieve paraphrase pool and training data pool
+    paraphrase_pool = create_paraphrase_pool(
+        paraphrases, pooled_intents, paraphrase_sim_score_threshold
+    )
+    (training_data_pool, training_data_vocab_per_intent,) = create_training_data_pool(
```

Comment:
This vocabulary per intent should be created inside max vocab expansion strategy function since it is specific to that strategy only.